multicluster/controllers/multicluster/member/clusterset_controller.go

```diff
@@ -130,17 +142,36 @@ func (r *MemberClusterSetReconciler) Reconcile(ctx context.Context, req ctrl.Req
 			err = r.Update(context.TODO(), clusterSet)
 			if err != nil {
 				klog.ErrorS(err, "Failed to update ClusterSet's ClusterID", "clusterset", req.NamespacedName)
-				return ctrl.Result{}, err
+				return err
 			}
 		}
-	}
-	r.clusterSetConfig = clusterSet.DeepCopy()
+		// ClusterSet deletion may fail and retry, but a ClusterSet maybe created just right before retry.
+		// ClusterSet deletion retry will become an update action, so try to delete stale resources here before
+		// initilize the new ClusterSet.
+		if r.clusterSetConfig != nil {
+			if err := r.deleteMemberClusterAnnounce(ctx); err != nil {
```

Comment:
Why not check whether cleanup is needed at the beginning of this func, like:
```
    if !clusterSetNotFound {
        // check if ClusterSet is recreated (by comparing ClusterID/ClusterSet ID and leader URL?), or updated or no change
        if noChange {
            return nil
        }
    }
    if clusterSetNotFound || clusterSetRecreated {
        // cleanup, stop common area, etc.
        if clusterSetNotFound {
            return nil
        }
    }
    // All other things like checking and setting ClusterSet ID in the spec, initializing commonArea and notifying stale controller..
```

Reply:
I am not very sure about leader URL. I feel if that changed, it should be treated the same as ClusterSet recreation, or we should even make it immutable too, but in this PR probably ok we keep the current logic and treat URL change as an update (so no cleanup, etc.)

Reply:
Yeah, I think we'd better to make the server URL immutable as well, otherwise, the scenarios will be complicated. I will refine this part in the future and also the secret update part. looks like there is an issue for existing implementation to handle the secret update from wrong one to correct one.